bazel: add patches to compile with gcc-11

glennpj opened the pull request:
Bazel does not currently build with GCC-11. There are patches available on the bazel project site for Alpine Linux that should work. The problem is that java-tools also needs patching.

glennpj committed:
bazel: new versions and patches to compile with gcc-11

adamjstewart commented:
Based on https://github.com/spack/spack/pull/27138#issuecomment-1019377748 I'm guessing that this wasn't enough to solve the issue?

adamjstewart commented on var/spack/repos/builtin/packages/bazel/package.py:
```diff
@@ -154,6 +158,19 @@ class Bazel(Package):
+    patch('gcc11_3.patch', when='@4.1:%gcc@11:')
+    patch('gcc11_4.patch', when='@4.1:%gcc@11:')
+
+    # bazel-4.0.0 does not compile with gcc-11
```
What about bazel 3? Should this be `@:3`?

glennpj replied:
I need to look at bazel 3 and see what needs to be done.

adamjstewart commented:
#29016 adds Bazel 5, might be worth looking at. If that fixes GCC 11 build issues I vote we close this.

glennpj committed:
Get older versions to build

- deprecate versions 0.16 and lower as they do not build
- versions older than 0.22 depend on java@8
- adjust when patches are applied
- adjust gcc conflicts directive

glennpj commented:
My apologies for this PR languishing so long. I committed another set of changes to address older versions.

1. Versions less than 4 can also be patched to get those versions to compile with newer GCC versions.
2. Version 4.0.0 is kind of a special case. As commented in the `package.py` file, it needs newer versions of grpc and abseil. I am not sure how to handle that. For now, that version is in a conflicts statement.


> #29016 adds Bazel 5, might be worth looking at. If that fixes GCC 11 build issues I vote we close this.

Bazel 5 does build without the patches. I allowed for that in the constraints settings of this PR.

adamjstewart commented:
Also see #30655

adamjstewart commented:
So, I need Bazel 4.1+ for Apple M1 support, and we have a few open Bazel PRs now, including #29016 and #30655. I'm no longer maintaining the Bazel package, which means that no one is maintaining the package, which means that no one cares if the package breaks. So I'm inclined to just start merging things and ask questions later. This PR seems to encompass most (all?) of #30655, and #29016 hasn't gotten a response since I reviewed it. So I'm thinking about just merging this and closing #30655 and #29016. Any objections from @elfprince13 or @haralmha? Any volunteers to officially maintain the package?

haralmha commented:
@adamjstewart sounds good to me. We are building the bazel based packages from wheels now anyway.

adamjstewart commented:
Tried building this on macOS 12.3.1 and Apple M1 Pro with Apple Clang 13.1.6 and JDK 18.0.1.1 but it doesn't build for me. Here is the build log: [spack-build-out.txt](https://github.com/spack/spack/files/8720566/spack-build-out.txt). Any idea why this doesn't work?

elfprince13 commented:
I can check tomorrow if this solves my issue.

elfprince13 commented:
Just tried building Bazel 4.2.2 and this also needs a patch for:

```console
  >> 72881    external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:26: error: no member named 'numeric_limits' in namespace 'std'
     72882      if (x->version == std::numeric_limits<uint32_t>::max()) {
     72883                        ~~~~~^
  >> 72884    external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:41: error: unexpected type name 'uint32_t': expected expression
     72885      if (x->version == std::numeric_limits<uint32_t>::max()) {
     72886                                            ^
  >> 72887    external/com_google_absl/absl/synchronization/internal/graphcycles.cc:451:52: error: no member named 'max' in the global namespace
     72888      if (x->version == std::numeric_limits<uint32_t>::max()) {
     72889                                                     ~~^
     72890    3 errors generated.
     72891    [935 / 2,844] 7 actions running
     72892        JavacBootstrap .../devtools/build/buildjar/libstarlark-deps.jar; 48s local
     72893        Compiling src/google/protobuf/descriptor.cc; 12s local
```

glennpj commented:
> Just tried building Bazel 4.2.2 and this also needs a patch for:

@elfprince13 This PR patches that already. Version 4.2.2 builds fine for me.

adamjstewart commented:
@glennpj when you get a chance can you rebase and make sure your patches still apply to all versions? @elfprince13 and I can test this and get it merged soon.

adamjstewart commented on var/spack/repos/builtin/packages/bazel/package.py:
```diff
@@ -20,14 +20,18 @@ class Bazel(Package):
+    version('5.1.1',  sha256='7f5d3bc1d344692b2400f3765fd4b5c0b636eb4e7a8a7b17923095c7b56a4f78')
+    version('5.1.0',  sha256='4de301f509fc6d0cbc697b2017384ecdc94df8f36245bbcbedc7ea6780acc9f5')
+    version('5.0.0',  sha256='072dd62d237dbc11e0bac02e118d8c2db4d0ba3ba09f1a0eb1e2a460fb8419db')
+    version('4.2.2',  sha256='9981d0d53a356c4e87962847750a97c9e8054e460854748006c80f0d7e2b2d33')
+    version('4.2.1',  sha256='12ea7aa11e2bdb12de1dceb9939a22e96f5a480437cb17c123379d8e0fdf5e82')
+    version('4.2.0',  sha256='74814b63920aaee47dbbbee7082e5c4317e4eebaf07e03c5fb5626e1716f1034')
+    version('4.1.0',  sha256='f377d755c96a50f6bd2f423562598d822f43356783330a0b780ad442864d6eeb')
+    version('4.2.2',  sha256='9981d0d53a356c4e87962847750a97c9e8054e460854748006c80f0d7e2b2d33')
+    version('4.2.1',  sha256='12ea7aa11e2bdb12de1dceb9939a22e96f5a480437cb17c123379d8e0fdf5e82')
+    version('4.2.0',  sha256='74814b63920aaee47dbbbee7082e5c4317e4eebaf07e03c5fb5626e1716f1034')
+    version('4.1.0',  sha256='f377d755c96a50f6bd2f423562598d822f43356783330a0b780ad442864d6eeb')
```
Some of these versions are duplicated, bad rebase? I would also avoid adding the extra space for alignment, we're about to blacken all of Spack so this will be undone anyway.

adamjstewart commented on var/spack/repos/builtin/packages/bazel/package.py:
```diff
@@ -80,43 +84,117 @@ class Bazel(Package):
-    version('0.1.1',  sha256='c6ae19610b936a0aa940b44a3626d6e660fc457a8187d295cdf0b21169453d20')
+    version('0.16.1',
+            sha256='09c66b94356c82c52f212af52a81ac28eb06de1313755a2f23eeef84d167b36c',
+            deprecated=True)
```
Not opposed to deprecating older versions, but why this specific set of versions? This also effectively deprecates TF 1.12 and older.

adamjstewart commented:
Need to rebase/merge master. Probably want to run `@spackbot fix style` before that though.

adamjstewart commented:
Oops, didn't intend to run that for you...

glennpj commented:
@spackbot fix style

glennpj committed:
[@spackbot] updating style on behalf of glennpj

glennpj commented:
So, spackbot adds a `conflicts` label but does not remove it when conflicts are resolved.

adamjstewart commented:
The conflicts label is for usage of the conflicts directive in a package, not for merge conflicts. Maybe we should remove that label, I agree that it's kind of confusing.

glennpj commented:
> The conflicts label is for usage of the conflicts directive in a package, not for merge conflicts. Maybe we should remove that label, I agree that it's kind of confusing.

Referring to package conflicts makes perfect sense, but my brain did associate it with the merge conflicts at the time. Perhaps `package-conflicts` or something similar.

adamjstewart commented:
Is this PR good to go now? @elfprince13 can you test this and make sure it solves your problems?

adamjstewart committed:
Ensure that patches apply to all versions

adamjstewart reviewed (approved):
Added a commit to fix some issues I noticed during patching. With these changes, all patches now apply successfully to all versions. I haven't actually tested building older bazel with newer GCC, but I'll leave that part to someone else to test.

adamjstewart commented on var/spack/repos/builtin/packages/bazel/gcc11_2.patch:
```diff
@@ -0,0 +1,14 @@
+diff --git a/third_party/ijar/zlib_client.h b/third_party/ijar/zlib_client.h
```
Any reason not to combine 1 and 2 into a single patch?

glennpj replied:
No, the reason they are separate was just to be consistent with the source the patches were derived from. That is also why they are named the way they are. I thought that would be the most clear way to track what is going on.

aweits commented:
> Added a commit to fix some issues I noticed during patching. With these changes, all patches now apply successfully to all versions. I haven't actually tested building older bazel with newer GCC, but I'll leave that part to someone else to test.

works over here....

```
[aweits@skl-a-00 spack]$ spack find bazel@4.1.0 %gcc@11.2.0
==> 1 installed package
-- linux-rhel7-skylake_avx512 / gcc@11.2.0 ----------------------
bazel@4.1.0
```
